### ISSUE TITLE
macOS 12 Monterey and macOS 11 Big Sur 

### ISSUE THREAD

**Issue by myspaghetti** [myspaghetti/macos-virtualbox#547]
The issue with installing macOS 12 Monterey is almost identical to [issue #254](https://github.com/myspaghetti/macos-virtualbox/issues/254#issuecomment-750889854) regarding installing macOS 11 Big Sur.

Apple packaged the file structure of the bootable installer in a `xar` archive (more accurately a `pkg` archive), so unpacking the installer for Big Sur and Monterey requires a `xar`-supporting archiver. This is available on macOS but rarely available as a binary package on other operating systems. [Compiling the free and open-source `xar` is not difficult](https://github.com/cylance/macos-arm64-emulation), but requiring compilation will impose many more dependencies on the script.

People familiar with compiling software can easily compile `xar` and use it to create a bootable installer from `InstallAssistant.pkg` the same way the script does. Meanwhile people who don't want to compile software can use the script to install earlier versions of macOS and use Software Update to create bootable installers of Monterey and Big Sur.

@rhoehener [suggested using an existing macOS VM to create a Big Sur bootable installer](https://github.com/myspaghetti/macos-virtualbox/issues/254#issuecomment-857531061), which would work all the same on Monterey and as an upgrade as well as a bootable installer.

This would work, and I think it's better to have it as a separate script which can be daisy-chained from this script if desired or applied to an existing macOS VM.

Currently I don't plan on implementing this, but it shouldn't be more or less straightforward to use `tesseract` similar to how it's used in the current script to detect the macOS VM state according to on-screen text and send the appropriate keystrokes.

Others have suggested using 7zip; it does not correctly extract this type of `xar` archives. Others have suggested using Docker; Docker on Windows uses Hyper-V and in general cannot be run concurrently with a VirtualBox macOS guest.

I deleted the previous issue (#476) to avoid the Google results that lead people to post about general issues with Docker and macOS here.

**Comment by haphaeu** [comment]
I'm willing to give it a try to a `xar` dependent fork of this project, for the people that want to compile `xar` themselves. Unless someone has already done that?

I'm however stuck where to find URLs for the [sucatalogs](https://github.com/myspaghetti/macos-virtualbox/blob/ce3ee4c410b7d46a46cb8cc2ab380b97fab5edee/macos-guest-virtualbox.sh#L394) for Big Sur and Monterey?

**Comment by myspaghetti** [comment]
Simple - they're found on working systems with Big Sur and Monterey! 😉

Big Sur software update catalog URL: `https://swscan.apple.com/content/catalogs/others/index-11-10.15-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog`

Monterey software update catalog URL: `https://swscan.apple.com/content/catalogs/others/index-12-10.16-10.15-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog`

**Comment by thispsj** [comment]
Hi @myspaghetti ,

I read in #254 about `7zip`. You said that it's only able to extract a single payload file.

I just wanted to let you know that 7zip can also extract the payload. I have attached a clip below for the same. Let me know if I am wrong 😅 

[!Payload Extraction](https://user-images.githubusercontent.com/51746608/156708162-77f363db-8e8a-430c-b8bc-ead3908f12e3.mp4)

**Comment by myspaghetti** [comment]
Oh nice, I never thought of trying to extract `Payload~`

**Comment by thispsj** [comment]
> Oh nice, I never thought of trying to extract `Payload~`

I know this from the time I installed Big Sur the last year but had never read #254 :sweat_smile: .

**Comment by thispsj** [comment]
> I initially ran into an infinite boot loop on upgrading from Catalina to Monterey using System Update. This solution on #484 worked for me courtesy of [FlorianLeMenn](https://github.com/FlorianLeMenn): [#484 (comment)](https://github.com/myspaghetti/macos-virtualbox/issues/484#issuecomment-1074848640).
> 
> TLDR in the host, with the macOS vm off:
> 
> ```shell
> VBoxManage setextradata "macos" "VBoxInternal/Devices/smc/0/Config/GetKeyFromRealSMC" 1
> VBoxManage setextradata "macos" "VBoxInternal/TM/TSCMode" "RealTSCOffset"
> ```
> 
> You don't need a fresh Catalina install to fix the boot loop. Stop the VM, run those commands, and it should complete the upgrade.
> 
> Note that the script sets `GetRealFromSMC` to `0` https://github.com/myspaghetti/macos-virtualbox/blob/master/macos-guest-virtualbox.sh#L753. It must remain `1` after Monterey is installed.

@nmonti you're absolutely right. Even I was facing the same issue but it took me 2 days before I could figure out the problem.

Putting this here will ensure that others don't have to waste so much of time like me😅

**Comment by R3D347HR4Y** [comment]
@nmonti Your solution worked and Monterey is currently working on my VM, there is one issue however:
My Display memory is at 7MB and graphical performance is insanely slow, tried installing VBox GAs and the install failed...
Tried with the VMWare ones and it doesn't change anything from what I can see...
I'll dial down the resolution for the time being but I'd like to know if someone found a solution to get better graphical performance on Monterey
![image](https://user-images.githubusercontent.com/28823392/167100290-48895ad8-ac2d-4393-a75f-5d754f461400.png)

**Comment by nmonti** [comment]
From my understanding macOS uses way more graphical power than other OSes. In addition, VMs aren't great with graphics and GPU passthrough is still seemingly in infancy. I was fiddling around and couldnt get it to be smooth despite jacking my VM settings up, but I'd love to know if there are any tricks!

**Comment by myspaghetti** [comment]
The best solution by far is to use QEMU/KVM instead of VirtualBox.

**Comment by leiless** [comment]
> 

Hi, @myspaghetti. How can I run macOS 11/12 in QEMU/KVM under a Linux host? is there any documentation?

**Comment by myspaghetti** [comment]
There's [a very nice tutorial by Nicholas Sherlock](https://www.nicksherlock.com/2021/10/installing-macos-12-monterey-on-proxmox-7/). You can export an existing VirtualBox image instead of creating a new installation. You can install OpenCore on the existing image or on a separate image. QEMU/KVM also works surprisingly well on [Windows through WSL2 or WSLg](https://boxofcables.dev/accelerated-kvm-guests-on-wsl-2/), you don't even need to compile any kernel modules, it works out of the box (test with `kvm-ok`).

**Comment by mr-gosh** [comment]
I just did the updates via system upgrade up to Monterey - worked like a charm.

Nit testing Ventura now.

**Comment by duyleekun** [comment]
The script works with minor modifications to install from internet recovery. You can always duplicate/snapshot machines so keeping the file for reusing later won't be useful I suppose.

**Comment by JayCynine** [comment]
Solution
To resolve the issues with installing macOS 12 Monterey and macOS 11 Big Sur, we need a xar-supporting archiver that can extract the files from the InstallAssistant.pkg file. Here are some possible solutions:

Solution 1: Use an Existing macOS VM
One solution is to use an existing macOS VM to create a bootable installer. This is a reliable and straightforward method, but it requires access to an existing macOS VM. Here's how we can create a bootable installer using an existing macOS VM:

Install macOS on a virtual machine using a hypervisor like VirtualBox or VMware.

Download the InstallAssistant.pkg file for Monterey or Big Sur.

Run the InstallAssistant.pkg file on the macOS VM to create a bootable installer.

Copy the bootable installer to a USB drive.

Here's some example code to automate this process using the VBoxManage command-line tool:
# Start the macOS VM
VBoxManage startvm "macOS VM" --type headless

# Wait for the VM to boot up
sleep 60

# Copy the InstallAssistant.pkg file to the VM
VBoxManage guestcontrol "macOS VM" copyto InstallAssistant.pkg "/Users/Shared/"

# Install the bootable installer on the VM
VBoxManage guestcontrol "macOS VM" run --exe "/usr/sbin/installer" --args "-pkg" "--/Users/Shared/InstallAssistant.pkg" "-target" "/"

# Copy the bootable installer to a USB drive
diskutil list
diskutil unmountDisk /dev/disk2
sudo dd if=/dev/disk2 of=monterey_usb.img bs=1m

Solution 2: Compile and Use xar
Another solution is to compile the free and open-source xar tool and use it to extract the files from the InstallAssistant.pkg file. This requires compiling software, but it can be done on most platforms. Here's how we can extract the files using xar:

Download and compile the xar tool.

Extract the contents of the InstallAssistant.pkg file using xar.

Copy the extracted files to a USB drive.

Here's some example code to automate this process:
# Download and compile xar
git clone https://github.com/mackyle/xar.git
cd xar
./autogen.sh
./configure
make
sudo make install

# Extract the contents of the InstallAssistant.pkg file
xar -xf InstallAssistant.pkg
cd InstallAssistant.pkg/Contents/SharedSupport/

# Copy the extracted files to a USB drive
diskutil list
diskutil unmountDisk /dev/disk2
sudo dd if=macOS\ Install\ Data/BaseSystem/BaseSystem.dmg of=/dev/disk2 bs=1m
Solution 3: Use an Alternative Archiver
There are several alternative archivers that can extract the files from the InstallAssistant.pkg file, such as pkgutil and unar. However, these may not work correctly on all systems or with all versions of the installer.

Solution 4: Use a Virtual Machine in Docker
One possible solution is to use a virtual machine in Docker to create the bootable installer. This allows us to use macOS without installing it on bare metal hardware. Here's how we can create a bootable installer using Docker:

Install Docker on the host system.

Download the InstallAssistant.pkg file for Monterey or Big Sur.

Create a Docker image with macOS installed using a tool like macOS-Simple-KVM.

Run the Docker image and mount the InstallAssistant.pkg file as a volume.
Install the bootable installer on the Docker image.

Copy the bootable installer to a USB drive.

Here's some example code to automate this process:

bash
# Download the InstallAssistant.pkg file
curl -O https://example.com/InstallAssistant.pkg

# Create a Docker image with macOS installed
git clone https://github.com/foxlet/macOS-Simple-KVM.git
cd macOS-Simple-KVM
./jumpstart.sh --mojave --cpus 2 --mem 4096

# Run the Docker image and mount the InstallAssistant.pkg file as a volume
docker run -it --rm -v $(pwd)/InstallAssistant.pkg:/InstallAssistant.pkg --privileged --device /dev/kvm -p 5900:5900 macos

# Install the bootable installer on the Docker image
cd /Volumes/Install\ macOS\ Monterey/
sudo ./createinstallmedia --volume /Volumes/MyVolume --nointeraction

# Copy the bootable installer to a USB drive
diskutil list
diskutil unmountDisk /dev/disk2
sudo dd if=/dev/disk2 of=monterey_usb.img bs=1m

Note that this method requires a Docker image with macOS installed, which can be created using a tool like macOS-Simple-KVM. Also, running Docker with privileged access and device access to /dev/kvm can pose security risks, so it should only be done in a secure environment.

**Comment by Tim-Cao** [comment]
VBoxManage setextradata "macos" "VBoxInternal/Devices/smc/0/Config/GetKeyFromRealSMC" 1
VBoxManage setextradata "macos" "VBoxInternal/TM/TSCMode" "RealTSCOffset"
This way doesn't work with upgrade to Ventura
![image](https://github.com/myspaghetti/macos-virtualbox/assets/52661397/e96bfabc-f2fc-4f4a-8ec2-605ec9e44864)

**Comment by BrianHotopp** [comment]
> > I initially ran into an infinite boot loop on upgrading from Catalina to Monterey using System Update. This solution on #484 worked for me courtesy of [FlorianLeMenn](https://github.com/FlorianLeMenn): [#484 (comment)](https://github.com/myspaghetti/macos-virtualbox/issues/484#issuecomment-1074848640).
> > TLDR in the host, with the macOS vm off:
> > ```shell
> > VBoxManage setextradata "macos" "VBoxInternal/Devices/smc/0/Config/GetKeyFromRealSMC" 1
> > VBoxManage setextradata "macos" "VBoxInternal/TM/TSCMode" "RealTSCOffset"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You don't need a fresh Catalina install to fix the boot loop. Stop the VM, run those commands, and it should complete the upgrade.
> > Note that the script sets `GetRealFromSMC` to `0` https://github.com/myspaghetti/macos-virtualbox/blob/master/macos-guest-virtualbox.sh#L753. It must remain `1` after Monterey is installed.
> 
> @nmonti you're absolutely right. Even I was facing the same issue but it took me 2 days before I could figure out the problem.
> 
> Putting this here will ensure that others don't have to waste so much of time like me😅

legend - this worked for me
